presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java

```diff
@@ -0,0 +1,513 @@
+        return serialize(leftGeometry.difference(rightGeometry));
+    }
+
+    @Description("For geometry type Returns the 2-dimensional cartesian minimum distance (based on spatial ref) between two geometries in projected units")
+    @ScalarFunction("ST_Distance")
+    @SqlType(StandardTypes.DOUBLE)
+    public static double stDistance(@SqlType(NAME) Slice left, @SqlType(NAME) Slice right)
+    {
+        OGCGeometry leftGeometry = deserialize(left);
+        OGCGeometry rightGeometry = deserialize(right);
+        return leftGeometry.distance(rightGeometry);
+    }
+
+    @Description("Returns a line string representing the exterior ring of the POLYGON geometry")
+    @ScalarFunction("ST_ExteriorRing")
+    @SqlType(NAME)
+    public static Slice stExteriorRing(@SqlType(NAME) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(deserialize(input), "ST_ExteriorRing", POLYGON);
+        try {
+            Method method = OGCPolygon.class.getMethod("exteriorRing");
+            OGCLineString exteriorRing = (OGCLineString) method.invoke(geometry);
+            return serialize(exteriorRing);
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Description("Returns a geometry that represents the shared portion of geometry A and geometry B")
+    @ScalarFunction("ST_Intersection")
+    @SqlType(NAME)
+    public static Slice stIntersection(@SqlType(NAME) Slice left, @SqlType(NAME) Slice right)
+    {
+        OGCGeometry leftGeometry = deserialize(left);
+        OGCGeometry rightGeometry = deserialize(right);
+        return serialize(leftGeometry.intersection(rightGeometry));
+    }
+
+    @Description("Returns a geometry that represents the portions of A and B that do not intersect")
+    @ScalarFunction("ST_SymDifference")
+    @SqlType(NAME)
+    public static Slice stSymmetricDifference(@SqlType(NAME) Slice left, @SqlType(NAME) Slice right)
+    {
+        OGCGeometry leftGeometry = deserialize(left);
+        OGCGeometry rightGeometry = deserialize(right);
+        return serialize(leftGeometry.symDifference(rightGeometry));
+    }
+
+    @SqlNullable
+    @Description("Returns TRUE if and only if no points of B lie in the exterior of A, and at least one point of the interior of B lies in the interior of A")
+    @ScalarFunction("ST_Contains")
+    @SqlType(StandardTypes.BOOLEAN)
+    public static Boolean stContains(@SqlType(NAME) Slice left, @SqlType(NAME) Slice right)
+    {
+        OGCGeometry leftGeometry = deserialize(left);
+        OGCGeometry rightGeometry = deserialize(right);
+        OperatorSimpleRelation containsOperator = OperatorContains.local();
+        return containsOperator.execute(leftGeometry.getEsriGeometry(), rightGeometry.getEsriGeometry(), leftGeometry.getEsriSpatialReference(), null);
+    }
+
+    @SqlNullable
+    @Description("Returns TRUE if the supplied geometries have some, but not all, interior points in common")
```

Comment:
This definition seems correct. However, when I called this function with a line and a point on the line, I got FALSE. I believe this function should return TRUE in such a case. Does it work like this for you?

This is what I tried:

assertQuery("select st_crosses(st_line('linestring(0 0, 1 1)'), st_point(0, 0))", "select true");

https://postgis.net/docs/ST_Crosses.html

Reply:
Hi @mbasmanova I think:
st_crosses(st_line('linestring(0 0, 1 1)'), st_point(0, 0))

should return FALSE

as Point(0 0) is one point in linestring(0 0, 1 1), it is ALL in the LineString. ST_Crosses requires Not All in.

What do you think?

Reply:
@zhenxiao , I think you are right. Thanks for clarifying.